app/services/supplementary-billing/reissue-invoice.service.js

```diff
@@ -0,0 +1,276 @@
+ * directly to the original source invoice.
+ *
+ * @param {module:BillingInvoiceModel} sourceInvoice The invoice to be reissued
+ * @param {module:BillingBatchModel} reissueBillingBatch The billing batch that the new invoices should belong to
+ *
+ * @returns {Object} dataToReturn Data that has been generated while reissuing the invoice
+ * @returns {Object[]} dataToReturn.billingInvoices Array of billing invoices
+ * @returns {Object[]} dataToReturn.billingInvoiceLicences Array of billing invoice licences
+ * @returns {Object[]} dataToReturn.transactions Array of transactions
+ */
+
+async function go (sourceInvoice, reissueBillingBatch) {
+  const dataToReturn = {
+    billingInvoices: [],
+    billingInvoiceLicences: [],
+    billingTransactions: []
+  }
+
+  // When a reissue request is sent to the Charging Module, it creates 2 new invoices (one to cancel out the original
+  // invoice and one to be the new version of it) and returns their ids
+  const chargingModuleReissueResponse = await _sendReissueRequest(
+    reissueBillingBatch.externalId,
+    sourceInvoice.externalId
+  )
+
+  for (const chargingModuleReissueInvoiceInfo of chargingModuleReissueResponse) {
+    // Because we only have the CM invoice's id we now need to fetch its details via the "view invoice" endpoint
+    const chargingModuleReissueInvoice = await _sendViewInvoiceRequest(
+      reissueBillingBatch,
+      chargingModuleReissueInvoiceInfo
+    )
+
+    const reissueBillingInvoice = _retrieveOrGenerateBillingInvoice(
+      dataToReturn,
+      sourceInvoice,
+      reissueBillingBatch,
+      chargingModuleReissueInvoice
+    )
+
+    // The invoice we want to reissue will have one or more billing invoice licences on it which we need to re-create
+    for (const sourceInvoiceLicence of sourceInvoice.billingInvoiceLicences) {
+      const reissueInvoiceLicence = _retrieveOrGenerateBillingInvoiceLicence(
+        dataToReturn,
+        sourceInvoice,
+        reissueBillingInvoice.billingInvoiceId,
+        sourceInvoiceLicence
+      )
+
+      const chargingModuleLicence = _retrieveChargingModuleLicence(
+        chargingModuleReissueInvoice,
+        sourceInvoiceLicence.licenceRef
+      )
+
+      // The invoice licence we are re-creating will have one or more transactions on it which we also need to re-craete
+      for (const sourceTransaction of sourceInvoiceLicence.billingTransactions) {
+        // Get the original transaction from the charging module data so we can create our new one
+        const chargingModuleReissueTransaction = _retrieveChargingModuleTransaction(
+          chargingModuleLicence,
+          sourceTransaction.externalId
+        )
+
+        // We need to know if this is an invoice which is cancelling out the original so we can create our transactions
+        // accordingly (ie. creating credit transactions to cancel out debits and vice-versa)
+        const isCancellingInvoice = _isCancellingInvoice(chargingModuleReissueInvoiceInfo)
```

Comment:
Fortunately, the Charging Module has our back on this! When it creates the 2 invoices it works out creating a debit for a credit and vice-versa. So, premise is we need to 'copy' what the charging module has created into our DB but fill in the blanks for anything it doesn't provide.

The legacy code appears to have gone with an approach of

- start with the source transaction
- overwrite matching properties from the CM transaction
- strip out the ID

We can follow the same or go with starting with the CM transaction and filling in the blanks from the source. Either way, `isCredit` is sorted for us so we don't need to flip anything.